hail/src/main/scala/is/hail/expr/ir/analyses/SemanticHash.scala

```diff
@@ -0,0 +1,409 @@
+      case _: ArrayLen |
+           _: ArrayRef |
+           _: ArraySlice |
+           _: ArraySort |
+           _: ArrayZeros |
+           _: Begin |
+           _: BlockMatrixCollect |
+           _: CastToArray |
+           _: Coalesce |
+           _: CollectDistributedArray |
+           _: ConsoleLog |
+           _: Consume |
+           _: Die |
+           _: GroupByKey |
+           _: If |
+           _: InsertFields |
+           _: IsNA |
+           _: Let |
+           _: LiftMeOut |
+           _: MakeArray |
+           _: MakeNDArray |
+           _: MakeStream |
+           _: MakeStruct |
+           _: MatrixAggregate |
+           _: MatrixColsTable |
+           _: MatrixCount |
+           _: MatrixMapGlobals |
+           _: MatrixLiteral |
+           _: MatrixMapRows |
+           _: MatrixFilterRows |
+           _: MatrixMapCols |
+           _: MatrixFilterCols |
+           _: MatrixMapEntries |
+           _: MatrixFilterEntries |
+           _: MatrixDistinctByRow |
+           _: NDArrayShape |
+           _: NDArraySlice |
+           _: NDArrayReshape |
+           _: NDArrayWrite |
+           _: RelationalLet |
+           _: RNGSplit |
+           _: RNGStateLiteral |
+           _: StreamAgg |
+           _: StreamDrop |
+           _: StreamDropWhile |
+           _: StreamFilter |
+           _: StreamFlatMap |
+           _: StreamFold |
+           _: StreamFold2 |
+           _: StreamFor |
+           _: StreamIota |
+           _: StreamLen |
+           _: StreamMap |
+           _: StreamRange |
+           _: StreamTake |
+           _: StreamTakeWhile |
+           _: TableGetGlobals |
+           _: TableAggregate |
+           _: TableCollect |
+           _: TableCount |
+           _: TableDistinct |
+           _: TableFilter |
+           _: TableMapGlobals |
+           _: TableMapRows |
+           _: TableRename |
+           _: ToArray |
+           _: ToDict |
+           _: ToSet |
+           _: ToStream |
+           _: Trap =>
+        ()
+
+      // Discrete values
+      case _: Void | _: True | _: False =>
+        ()
+
+      // integral constants
+      case I32(x) => buffer ++= Bytes.fromInt(x)
+      case I64(x) => buffer ++= Bytes.fromLong(x)
+      case F32(x) => buffer ++= Bytes.fromFloat(x)
+      case F64(x) => buffer ++= Bytes.fromDouble(x)
+      case NA(typ) => buffer ++= SemanticTypeName(typ)
+      case Str(x) => buffer ++= x.getBytes()
+
+      // In these cases, just return None meaning that two
+      // invocations will never return the same thing.
+      case _ =>
+        log.warn(s"SemanticHash unknown: ${ir.getClass.getName}")
+        return None
+    }
+
+    Some(buffer.result())
+  }
+
+  def getFileHash(fs: FS)(path: String): Array[Byte] =
+    fs.eTag(path) match {
+      case Some(etag) =>
+        etag.getBytes
+      case None =>
+        path.getBytes ++ Bytes.fromLong(fs.fileStatus(path).getModificationTime)
+    }
```

Comment:
Something to call out here:
I'm just using the file's etag if the filesystem supports them and NOT the path. 
I think that the etag is unique on azure (their doc is quite hard to navigate so I'm finding it hard to know for sure - welcome help!).
GCS has the nice property that copying the file preserves the etag and it only changes after modification.